File: plotly_62/interactiveplot_cpp/KPI/c_business_layer/kpi_factory.cpp
#include "kpi_factory.h"
#include "alignment_matching_kpi.h"
#include "detection_matching_kpi.h"
#include "tracker_matching_kpi.h"
#include "../b_data_storage/kpi_config_storage.h"
#include <iostream>
#include <filesystem>
#include <random>
#include <sstream>
#include <thread>
#include <future>

namespace kpi {
namespace business {

KpiDataModel::KpiDataModel(std::shared_ptr<data_storage::KPIDataModelStorage> input_data,
                           std::shared_ptr<data_storage::KPIDataModelStorage> output_data,
                           const std::string& sensor)
    : input_data_(input_data), output_data_(output_data), sensor_(sensor) {
    
    temp_dir_ = create_temp_directory();
    process_kpis();
}

KpiDataModel::~KpiDataModel() {
    cleanup_temp_directory();
}

void KpiDataModel::process_kpis() {
    std::cout << "Processing KPIs in parallel using multithreading" << std::endl;
    
    // Launch KPI processing tasks in parallel
    std::vector<std::future<void>> futures;
    
    futures.push_back(std::async(std::launch::async, [this]() {
        process_alignment_kpis();
    }));
    
    futures.push_back(std::async(std::launch::async, [this]() {
        process_detection_kpis();
    }));
    
    futures.push_back(std::async(std::launch::async, [this]() {
        process_tracker_kpis();
    }));
    
    // Wait for all tasks to complete
    for (auto& future : futures) {
        future.wait();
    }
    
    std::cout << "All KPI processing completed" << std::endl;
}

void KpiDataModel::process_alignment_kpis() {
    auto input_data = input_data_->get_data();
    auto output_data = output_data_->get_data();
    
    for (const auto& config_pair : data_storage::KPI_ALIGNMENT_CONFIG) {
        const std::string& stream_name = config_pair.first;
        
        // Check if stream exists in both input and output data
        if (input_data.find(stream_name) != input_data.end() && 
            output_data.find(stream_name) != output_data.end()) {
            
            std::cout << "Processing alignment KPI for stream: " << stream_name << std::endl;
            
            try {
                std::string plot_path = process_alignment_kpi(*input_data_, *output_data_, 
                                                           stream_name, temp_dir_);
                if (!plot_path.empty()) {
                    plots_.push_back(plot_path);
                }
            } catch (const std::exception& e) {
                std::cerr << "Error processing alignment KPI for " << stream_name 
                         << ": " << e.what() << std::endl;
            }
        }
    }
}

void KpiDataModel::process_detection_kpis() {
    auto input_data = input_data_->get_data();
    auto output_data = output_data_->get_data();
    
    for (const auto& config_pair : data_storage::KPI_DETECTION_CONFIG) {
        const std::string& stream_name = config_pair.first;
        
        // Check if stream exists in both input and output data
        if (input_data.find(stream_name) != input_data.end() && 
            output_data.find(stream_name) != output_data.end()) {
            
            std::cout << "Processing detection KPI for stream: " << stream_name << std::endl;
            
            try {
                std::string plot_path = process_detection_kpi(*input_data_, *output_data_, 
                                                            stream_name, temp_dir_);
                if (!plot_path.empty()) {
                    plots_.push_back(plot_path);
                }
            } catch (const std::exception& e) {
                std::cerr << "Error processing detection KPI for " << stream_name 
                         << ": " << e.what() << std::endl;
            }
        }
    }
}

void KpiDataModel::process_tracker_kpis() {
    auto input_data = input_data_->get_data();
    auto output_data = output_data_->get_data();
    
    for (const auto& config_pair : data_storage::KPI_TRACKER_CONFIG) {
        const std::string& stream_name = config_pair.first;
        
        // Check if stream exists in both input and output data
        if (input_data.find(stream_name) != input_data.end() && 
            output_data.find(stream_name) != output_data.end()) {
            
            std::cout << "Processing tracker KPI for stream: " << stream_name << std::endl;
            
            try {
                std::string plot_path = process_tracker_kpi(*input_data_, *output_data_, 
                                                          stream_name, temp_dir_);
                if (!plot_path.empty()) {
                    plots_.push_back(plot_path);
                }
            } catch (const std::exception& e) {
                std::cerr << "Error processing tracker KPI for " << stream_name 
                         << ": " << e.what() << std::endl;
            }
        }
    }
}

std::string KpiDataModel::create_temp_directory() {
    // Generate random directory name
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1000, 9999);
    
    std::stringstream ss;
    ss << "kpi_plots_temp_" << dis(gen);
    
    std::filesystem::path temp_path = std::filesystem::temp_directory_path() / ss.str();
    std::filesystem::create_directories(temp_path);
    
    return temp_path.string();
}

void KpiDataModel::cleanup_temp_directory() {
    if (!temp_dir_.empty() && std::filesystem::exists(temp_dir_)) {
        try {
            std::filesystem::remove_all(temp_dir_);
        } catch (const std::exception& e) {
            std::cerr << "Error cleaning up temp directory: " << e.what() << std::endl;
        }
    }
}

std::vector<std::string> KpiDataModel::get_plots() const {
    return plots_;
}

std::string KpiDataModel::get_temp_dir() const {
    return temp_dir_;
}

} // namespace business
} // namespace kpi

File: sadfdsaffdsasf/KPI/c_business_layer/kpi_factory.py
import os
import tempfile
import logging
import json
import gc
from typing import Optional, List, Dict, Any
from KPI.b_data_storage.kpi_data_model_storage import KPI_DataModelStorage
from KPI.c_business_layer.alignment_matching_kpi import process_alignment_kpi
from KPI.c_business_layer.detection_matching_kpi import process_detection_kpi
from KPI.c_business_layer.tracker_matching_kpi import process_tracker_kpi
import concurrent.futures

class KpiDataModel:
    """
    Class for generating KPI plots from KPI_DataModelStorage.
    This class processes data from KPI_DataModelStorage to generate KPI visualizations.
    """
    
    def __init__(self, data: Dict[str, KPI_DataModelStorage], sensor: str):
        """
        Initialize the KPI data model with input and output data.
        
        Parameters:
            data: Dictionary containing 'input' and 'output' KPI_DataModelStorage instances
            sensor: Name of the sensor being processed
        """
        self.data = data
        self.sensor = sensor

        # Create temporary directory for plot files
        self.temp_dir = tempfile.mkdtemp(prefix="kpi_plots_temp_")
        
        # Initialize plots list
        self.plots = []
        
        # Process KPIs based on stream type
        self._process_kpis()
    
    def _process_kpis(self):
        """
        Process KPIs in parallel using multiprocessing
        Runs alignment, tracker, and detection KPIs concurrently
        """
        import multiprocessing as mp
        
        # Determine which KPIs to process based on stream name
        tasks = [
            # ("alignment", self._process_alignment_kpi),
            # ("tracker", self._process_tracker_kpi),
            ("detection", self._process_detection_kpi)
        ]
            
        if not tasks:
            logging.warning(f"No KPI processors available for stream: {self.stream_name}")
            self.plots = []
            return
            
        # Process tasks in parallel
        self.plots = []
        with concurrent.futures.ProcessPoolExecutor(max_workers=len(tasks)) as executor:
            futures = {executor.submit(func): name for name, func in tasks}
            for future in concurrent.futures.as_completed(futures):
                try:
                    self.plots.extend(future.result())
                except Exception as e:
                    logging.error(f"Error processing {futures[future]} KPI: {e}")
    
    def _process_alignment_kpi(self):
        """
        Process alignment KPIs using the alignment_matching_kpi module.
        """
        try:
            from KPI.c_business_layer.alignment_matching_kpi import process_alignment_kpi
            
            # Process alignment KPI for each sensor
            results = []
            for sensor_id, stream_name in self._get_alignment_streams():
                result = process_alignment_kpi(
                    self.data, 
                    sensor_id,
                    stream_name
                )
                results.append(result)
            
            return results
        except Exception as e:
            logging.error(f"Error processing alignment KPIs: {str(e)}")
            return []

    def _process_detection_kpi(self):
        """
        Process detection KPIs using the detection_matching_kpi module.
        """
        try:
            from KPI.c_business_layer.detection_matching_kpi import process_detection_kpi
            
            # Process detection KPI for each sensor
            results = []
            result = process_detection_kpi(
                    self.data,
                    self.sensor,
                )
            results.append(result)
            
            return results
        except Exception as e:
            logging.error(f"Error processing detection KPIs: {str(e)}")
            return []

    def _process_tracker_kpi(self):
        """
        Process tracker KPIs using the tracker_matching_kpi module.
        """
        try:
            from tracker_matching_kpi import process_tracker_kpi
            
            # Process tracker KPI for each sensor
            results = []
            for sensor_id, stream_name in self._get_tracker_streams():
                result = process_tracker_kpi(
                    self.input_data,
                    self.output_data,
                    sensor_id,
                    stream_name
                )
                results.append(result)
            
            return results
        except Exception as e:
            logging.error(f"Error processing tracker KPIs: {str(e)}")
            return []

    def _get_alignment_streams(self):
        """Get alignment streams from config"""
        return [("FC", "alignment"), ("FL", "alignment"), ("FR", "alignment")]

    def _get_detection_streams(self):
        """Get detection streams from config"""
        return [("FC", "detection"), ("FL", "detection"), ("FR", "detection")]

    def _get_tracker_streams(self):
        """Get tracker streams from config"""
        return [("FC", "tracker"), ("FL", "tracker"), ("FR", "tracker")]
